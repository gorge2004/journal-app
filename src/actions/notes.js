import { db } from "../firabase/firebase-config";
import { types } from "../types/types";
import { loadNotes } from "../helpers/loadNotes";
import Swal from "sweetalert2";
import { fileUpload } from "../helpers/fileUpload";
import { type } from "@testing-library/user-event/dist/type";
//react-journal
export const startNewNote = () => {
  return async (dispatch, getState) => {
    //get current state on redux
    const uid = getState().auth.uid;
    const newNote = {
      title: "",
      body: "",
      date: new Date().getTime(),
    };
    try {
      const doc = await db.collection(`${uid}/journal/notes`).add(newNote);
      dispatch(activetNote(doc.id, newNote));
      dispatch(addNewNote(doc.id, newNote));
    } catch (error) {}
  };
};

export const activetNote = (id, note) => {
  return {
    type: types.notesActive,
    payload: {
      id: id,
      ...note,
    },
  };
};
export const addNewNote = (id, note) => ({
  type: types.notesAddNew,
  payload: { id, ...note },
});

export const startLoadingNotes = (uid) => {
  return async (dispatch) => {
    const notes = await loadNotes(uid);
    dispatch(setNotes(notes));
  };
};

export const setNotes = (notes) => {
  return {
    type: types.notesLoad,
    payload: notes,
  };
};

export const startSaveNote = (note) => {
  return async (dispatch, getState) => {
    for (const key in note) {
      if (note[key] === undefined) {
        delete note[key];
      }
    }
    try {
      const uid = getState().auth.uid;
      const noteToFirestore = { ...note };
      delete noteToFirestore.id;
      await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore);

      dispatch(refreshNote(note.id, noteToFirestore));

      Swal.fire("Saved", note.title, "success");
    } catch (error) {
    }
  };
};
export const refreshNote = (id, note) => ({
  type: types.notesUpdated,
  payload: { id, note: { id, ...note } },
});

export const startuploadingNote = (file) => {
  return async (dispatch, getState) => {
    const { active: activeNote } = getState().notes;
    Swal.fire({
      title: "Uploading...",
      text: "Please Wait...",
      allowOutsideClick: false,
      showConfirmButton: false,
      willOpen: () => {
        Swal.showLoading();
      },
    });
    const urlImage = await fileUpload(file);
    activeNote.url = urlImage;
    dispatch(startSaveNote(activeNote));
    Swal.close();
  };
};

export const startDeleting = (id) => {
  return async (dispatch, getState) => {
    const uid = await getState().auth.uid;
    await db.doc(`${uid}/journal/notes/${id}`).delete();
    dispatch(deleteNote(id));
  };
};

export const deleteNote = (id) => ({
  type: types.notesDelete,
  payload: id,
});

export const noteLogout = () => ({
  type: types.notesLogoutCleaning,
});
