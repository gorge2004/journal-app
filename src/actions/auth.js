import Swal from 'sweetalert2'
import { firebase, googleAuthProvider } from "../firabase/firebase-config";
import { types } from "../types/types";
import { noteLogout } from './notes';
import { finishLoading, removeError, setError, startLoading } from "./ui";

//el middleware va recibir una accion , aqui se regresa una  funcion
export const startLoginByEmailPassword = (email, password) => {
  return async (dispatch) => {
    dispatch(removeError());
    dispatch(startLoading());
   return  firebase
      .auth()
      .signInWithEmailAndPassword(email, password)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName));
        dispatch(finishLoading());
      })
      .catch((err) => {
        dispatch(setError(err.message));
        dispatch(finishLoading());
        Swal.fire('ERROR',err.message, 'error');
      });
  };
};
export const startGoogleLogin = () => {
  return (dispatch) => {
    firebase
      .auth()
      .signInWithPopup(googleAuthProvider)
      .then((userCredential) => {
        const { uid, displayName } = userCredential.user;
        dispatch(login(uid, displayName));
      });
  };
};

export const startRegisterWithEmailAndPassword = (email, password, name) => {
  return (dispatch) => {
    firebase
      .auth()
      .createUserWithEmailAndPassword(email, password)
      .then(async ({ user }) => {
        await user.updateProfile({ displayName: name });
        console.log(user.uid, user.displayName);
        dispatch(login(user.uid, user.displayName));
      })
      .catch((error) => {
        Swal.fire('ERROR',error.message, 'error');

      });
  };
};

export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName,
  },
});

export const startLogout =  (email, password) => {
  return async (dispatch) => {
    await firebase.auth().signOut();

    dispatch(logout());
    dispatch(noteLogout());
  };
};

export const logout = () => ({ type: types.logout});
