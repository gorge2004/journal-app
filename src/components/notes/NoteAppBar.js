import React, { useRef } from "react";
import { useDispatch, useSelector } from "react-redux";
import { startSaveNote, startuploadingNote } from "../../actions/notes";

const NoteAppBar = () => {
  const fileRef = useRef();
  const dispatch = useDispatch();
  const { active: note } = useSelector((state) => state.notes);
  const handleSaveClick = () => {
    dispatch(startSaveNote(note));
  };
  const handlePictureClick = (evt) => {
    fileRef.current.click();
  };
  const handleFileChange = (evt) => {
    const file = evt.target.files[0];
    if (file) {
      dispatch(startuploadingNote(file));
    }
  };
  return (
    <div className="notes__appbar">
      <span>15 de febrero 2022</span>
      <input
        type="file"
        name="file"
        style={{
          display: "none",
        }}
        ref={fileRef}
        onChange={handleFileChange}
      />
      <div>
        <button className="btn" onClick={handlePictureClick}>
          Picture
        </button>
        <button className="btn" onClick={handleSaveClick}>
          Save
        </button>
      </div>
    </div>
  );
};

export default NoteAppBar;
