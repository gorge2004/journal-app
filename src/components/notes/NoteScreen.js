import { useEffect, useRef } from "react";
import NoteAppBar from "./NoteAppBar";
import { useDispatch, useSelector } from "react-redux";
import { useForm } from "../../hooks/useForm";
import { activetNote, startDeleting } from "../../actions/notes";


function NoteScreen() {

  const { active: note } = useSelector((state) => state.notes);
  const dispatch = useDispatch();
  const [formValues, handleInputChange, reset] = useForm(note);
  const { id, title, body } = formValues;
  const activeId = useRef(note.id);
  useEffect(() => {
    if (note.id !== activeId.current) {
      reset(note);
      activeId.current = note.id;
    }
  }, [note, reset]);

  useEffect(() => {
   
    dispatch(activetNote(id,{... formValues}));
  }, [formValues, dispatch]);
  
  const handleDeleteClick = () => {
    dispatch(startDeleting(id));
  }

  return (
    <div className="notes__main-content">
      <NoteAppBar />
      <div className="notes__content">
        <input
          type="text"
          placeholder="Some awesome title"
          className="note__title-input"
          name="title"
          autoComplete="off"
          value={title}
          onChange={handleInputChange}
        />
        <textarea
          placeholder="What happened today"
          className="note__textarea"
          onChange={handleInputChange}
          name="body"
          value={body}
        ></textarea>

        {note.url && (
          <div className="notes__image">
            <img
              src={note.url }
              alt="imagen"
            />
          </div>
        )}
      </div>
      <button className="btn btn-danger" onClick={handleDeleteClick}>
        Delete
      </button>
    </div>
  );
}

export default NoteScreen;
