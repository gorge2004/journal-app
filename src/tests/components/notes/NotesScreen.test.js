
import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { activetNote } from "../../../actions/notes";
import NoteScreen from "../../../components/notes/NoteScreen";

jest.mock("../../../actions/auth", () => ({
  startLogout: jest.fn(),
}));
jest.mock("../../../actions/notes", () => ({
    activetNote: jest.fn(),
}));
const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const note =  {id: "123s", title: "John", body: "This is john"}
const initialState = {
  auth: { uid: "123", name: "John" },
  ui: {},
  notes: { notes: [note], active: note },
};
let store = mockStore(initialState);
store.dispatch = jest.fn(() => {});
const wrapper = mount(
  <Provider store={store}>
    <NoteScreen />
  </Provider>
);
describe('testing_noteScreenTest', () => { 

    test('should_render_NoteScreen_Component', () => {
        expect(wrapper).toMatchSnapshot();
    });
    test('should_call_ActivNote_Action', () => { 
        const title =  'hola de nuevo';
        wrapper.find('input[name="title"]').simulate('change',{
            target: {
                name:'title',
                value: title,
            }
        });

        expect(activetNote).toHaveBeenCalled();
        expect(activetNote).toBeCalledWith(note.id,{...note, title})
     });
 })