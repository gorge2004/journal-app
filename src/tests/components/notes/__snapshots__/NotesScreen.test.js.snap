// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing_noteScreenTest should_render_NoteScreen_Component 1`] = `
<div
  className="notes__main-content"
>
  <div
    className="notes__appbar"
  >
    <span>
      15 de febrero 2022
    </span>
    <input
      name="file"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        Picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      autoComplete="off"
      className="note__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Some awesome title"
      type="text"
      value="John"
    />
    <textarea
      className="note__textarea"
      name="body"
      onChange={[Function]}
      placeholder="What happened today"
      value="This is john"
    />
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
